File: vjezbe/vjezba_3/zd_5.cpp
#include <iostream>
#include <vector>
using namespace std;

int countSolutions = 0;

bool isSafe(const vector<int>& board, int row, int col) {
    for (int i = 0; i < row; ++i) {
        if (board[i] == col || board[i] - i == col - row || board[i] + i == col + row) {
            return false;
        }
    }
    return true;
}

void solveNQueens(vector<int>& board, int row, int n) {
    if (row == n) {
        countSolutions++;
        return;
    }

    for (int col = 0; col < n; ++col) {
        if (isSafe(board, row, col)) {
            board[row] = col;      
            solveNQueens(board, row + 1, n); 
        }
    }
}

int main() {
    int n;
    cin >> n;

    vector<int> board(n);
    solveNQueens(board, 0, n);

    cout << countSolutions << endl;
    return 0;
}

File: vjezbe/vjezba_3/zd_5.py
def is_safe(row, col, queens):
    for r, c in queens:
        if c == col or r - c == row - col or r + c == row + col:
            return False
    return True

def solve_n_queens(n, row=0, queens=[]):
    if row == n:
        return 1 
    count = 0
    for col in range(n):
        if is_safe(row, col, queens):
            count += solve_n_queens(n, row + 1, queens + [(row, col)])
    return count

n = int(input())
print(solve_n_queens(n))
